datalad/tests/fixtures.py

```diff
@@ -0,0 +1,155 @@
+"""utilities to assist with setting up tests"""
+
+from datalad import cfg
+from datalad.core.distributed.clone import (
+    Clone,
+    decode_source_spec
+)
+from datalad.distribution.dataset import Dataset
+from datalad.utils import (
+    better_wraps,
+    ensure_list,
+    optional_args,
+)
+from datalad.tests.utils import with_tempfile
+
+
+DATALAD_TESTS_CACHE = cfg.obtain("datalad.tests.cache")
+
+
+def get_cached_dataset(url, dataset_name=None, version=None, paths=None):
+    """ Helper to get a cached clone from url
+
+    Intended for use from within `cached_dataset` and `cached_url` decorators.
+    Clones `url` into user's cache under datalad/tests/`name`. If such a clone
+    already exists, don't clone but return the existing one. So, it's supposed
+    to cache the original source in order to reduce time and traffic for tests,
+    by letting subsequent requests clone from a local location directly.
+
+    If it's an annex get the content as provided by `paths`, too.
+
+    Verifies that `version` can be checked out, but doesn't actually do it,
+    since the cached dataset is intended to be used as origin instead of the
+    original remote at URL by the `cached_dataset` test decorator. Checkout of
+    a particular version should happen in its clone.
+
+    Parameters
+    ----------
+    url: str
+        URL to clone from
+    dataset_name: str or None
+        (directory) name to use for the clone. If None, a name will be derived
+        from `url`.
+    paths: str or list or None
+        (list of) paths to get content for. Passed to datalad-get.
+    version: str or None
+        A commit or an object that can be dereferenced to one.
+
+    Returns
+    -------
+    Dataset
+    """
+
+    # TODO: What about recursive? Might be complicated. We would need to make
+    #       sure we can recursively clone _from_ here then.
+
+    if not DATALAD_TESTS_CACHE:
+        raise ValueError("Caching disabled by config")
+
+    if not dataset_name:
+        dataset_name = decode_source_spec(url)['default_destpath']
+
+    ds = Dataset(DATALAD_TESTS_CACHE / dataset_name)
+    if not ds.is_installed():
+        ds = Clone()(url, ds.pathobj)
+    if version:
+        # check whether version is available
+        assert ds.repo.commit_exists(version)
+    # TODO: What about an outdated dataset in cache? How do we properly update
```

Comment:
I think an option isn't it. That's again because it is supposed to work as cache and thereby needs to be transparent to the test. If the cache was skipped, no update would ever be required. If we now need to specify an option on how to update the usage becomes messy, I think.
I'd rather do something like:
If the dataset exists in cache, fetch and just proceed if there was nothing to fetch. If there was something however, we would need to decide what to do with that (merge isn't necessarily it) or - simplest solution - throw cache away in that case and get a fresh one.

Reply:
Good point about transparency! Then update only if version isn't found locally or explicitly requested overall?
We also will benefit from an explicit trigger to reset or ignore cached version (relates to the "fresh one") - I had to do that in dandi as well (https://github.com/dandi/dandi-cli/blob/master/dandi/support/cache.py#L38) although for a different type of cache (see https://github.com/con/pyfscacher/issues/1 for a higher level description/discussion, forgot to ping you there). I can see how it would be useful here as well

Reply:
Will have a look. But quick remark:

> Then update only if version isn't found locally or explicitly requested overall?

Insufficient. "version" could be a branch. A test would "mean" the branch as available from `url`. But that one could have been updated. We would "find" `version` in an outdated dataset in cache.